registry/response/terraform_provider.go

```diff
@@ -0,0 +1,66 @@
+package response
+
+import (
+	"sort"
+
+	version "github.com/hashicorp/go-version"
+)
+
+// TerraformProvider is the response structure for all required information for
+// Terraform to choose a download URL. It must include all versions and all
+// platforms for Terraform to perform version and os/arch constraint matching
+// locally.
+type TerraformProvider struct {
+	ID       string `json:"id"`
+	Verified bool   `json:"verified"`
+
+	Versions []*TerraformProviderVersion `json:"versions"`
+}
+
+// TerraformProviderVersion is the Terraform-specific response structure for a
+// provider version.
+type TerraformProviderVersion struct {
+	Version   string   `json:"version"`
+	Protocols []string `json:"protocols"`
+
+	Platforms []*TerraformProviderPlatform `json:"platforms"`
+}
+
+// TerraformProviderVersions is the Terraform-specific response structure for an
+// array of provider versions
+type TerraformProviderVersions struct {
+	ID       string                      `json:"id"`
+	Versions []*TerraformProviderVersion `json:"versions"`
+}
+
+// TerraformProviderPlatform is the Terraform-specific response structure for a
+// provider platform.
+type TerraformProviderPlatform struct {
+	OS   string `json:"os"`
+	Arch string `json:"arch"`
+}
+
+// TerraformProviderPlatformLocation is the Terraform-specific response
+// structure for a provider platform with all details required to perform a
+// download.
+type TerraformProviderPlatformLocation struct {
+	OS                  string `json:"os"`
+	Arch                string `json:"arch"`
+	Filename            string `json:"filename"`
+	DownloadURL         string `json:"download_url"`
+	ShasumsURL          string `json:"shasums_url"`
+	ShasumsSignatureURL string `json:"shasums_signature_url"`
+}
+
+// Collection type for TerraformProviderVersion
+type Collection []*TerraformProviderVersion
```

Comment:
Given that this is specifically a collection of provider versions, should it perhaps be `ProviderVersionCollection`? That is, if we think there might be other "collection"-ish types in future for other reasons.

I think elsewhere we've used the convention of just pluralizing the thing being collected, like `type ProviderVersions []*ProviderVersion`, but that's subjective so I don't feel very strongly about it.

Reply:
That makes sense to me, and was also my initial instinct;  I copied the `Collection` pattern and name out of go-version.

Reply:
Ah yes, `TerraformProviderVersions` already exists in the registry package.

TBH I'm not confident that little bit of code even belongs here, though I think I should wait till a later phase of this work until I overthink that.

I keep thinking back to the `versions` pattern in `plugin/discovery` where you wrap `go-version` types and functions to keep imports to a small set of files and wondering if I should do the same with the registry response types.

Reply:
I'm honestly not sure what's best here with what we know so far, so I think it'd be fine to go ahead with `ProviderVersionCollection` for now and then we can look at it in a "big picture" sort of way in a later phase, when it's clearer where all of the parts are going to end up.